File: server/js/src/v2/history.ts
/**
 * Conversation-history handling: ad copy never becomes assistant "speech".
 * Mirrors monetzly/v2/history.py.
 */
import { Ad } from "./types.js";

const MARKER_RE = /⟦ad:[^⟧]*⟧[\s\S]*?⟦\/ad⟧/g;
const ANNOTATION_RE = /\[sponsored suggestion shown: ([^—\]]+) — ad:([^\]]+)\]/g;

export function annotationFor(ad: Ad): string {
  return `[sponsored suggestion shown: ${ad.brand} — ad:${ad.id}]`;
}

/** Replace any marker block with the annotation (or strip if no ad). */
export function rewriteAssistantText(rawText: string, ad: Ad | null): string {
  return rawText.replace(MARKER_RE, ad ? annotationFor(ad) : "");
}

function collapse(text: string): string {
  return text.toLowerCase().replace(/[^a-z0-9]/g, "");
}

const REFERRING_RE =
  /\b(that (product|tool|brand|service|offer|suggestion|company|one)|the sponsored|that ad|tell me more|more about (it|that|this)|what about (it|that)|is it (free|good|worth|expensive)|how much (is it|does it cost)|that one you (mentioned|suggested|recommended))\b/;

/**
 * Return the ad id the user appears to be asking about, if any.
 * Brand mentions match regardless of spacing/punctuation; vague referring
 * phrases only count when adIsRecent (ad shown within the last few turns).
 */
export function detectFollowup(
  userText: string,
  recentAds: Ad[],
  adIsRecent = true,
): string | null {
  if (!recentAds.length) return null;
  const collapsed = collapse(userText);
  for (const ad of [...recentAds].reverse()) {
    const brand = collapse(ad.brand);
    if (brand && collapsed.includes(brand)) return ad.id;
  }
  if (adIsRecent && REFERRING_RE.test(userText.toLowerCase())) {
    return recentAds[recentAds.length - 1]!.id;
  }
  return null;
}

/** Extract ad ids from annotations in stored history. */
export function findAnnotations(historyText: string): string[] {
  return [...historyText.matchAll(ANNOTATION_RE)].map((m) => m[2]!.trim());
}

File: server/python/monetzly/v2/history.py
"""
Conversation-history handling: the ad copy never becomes assistant "speech".
"""
import re
from typing import List, Optional

from .types import Ad

_ANNOTATION_TEMPLATE = "[sponsored suggestion shown: {brand} — ad:{ad_id}]"
_MARKER_RE = re.compile(r"⟦ad:[^⟧]*⟧.*?⟦/ad⟧", re.DOTALL)
_ANNOTATION_RE = re.compile(r"\[sponsored suggestion shown: (?P<brand>[^—\]]+) — ad:(?P<ad_id>[^\]]+)\]")


def annotation_for(ad: Ad) -> str:
    return _ANNOTATION_TEMPLATE.format(brand=ad.brand, ad_id=ad.id)


def rewrite_assistant_text(raw_text: str, ad: Optional[Ad]) -> str:
    """Replace any marker block with the annotation (or strip if no ad)."""
    replacement = annotation_for(ad) if ad else ""
    return _MARKER_RE.sub(replacement, raw_text)


def _collapse(text: str) -> str:
    return re.sub(r"[^a-z0-9]", "", text.lower())


_REFERRING_RE = re.compile(
    r"\b(that (product|tool|brand|service|offer|suggestion|company|one)|"
    r"the sponsored|that ad|tell me more|more about (it|that|this)|"
    r"what about (it|that)|is it (free|good|worth|expensive)|"
    r"how much (is it|does it cost)|"
    r"that one you (mentioned|suggested|recommended))\b")


def detect_followup(user_text: str, recent_ads: List[Ad],
                    ad_is_recent: bool = True) -> Optional[str]:
    """
    Return the ad_id the user appears to be asking about, if any.

    Brand mentions match regardless of spacing/punctuation ("go green energy
    solar" matches brand "GreenEnergy Solar"). Vague referring phrases
    ("tell me more about it") only count when ad_is_recent — the caller knows
    whether an ad was shown in the last few turns.
    """
    if not recent_ads:
        return None
    lowered = user_text.lower()
    collapsed = _collapse(user_text)
    for ad in reversed(recent_ads):
        brand_collapsed = _collapse(ad.brand)
        if brand_collapsed and brand_collapsed in collapsed:
            return ad.id
    if ad_is_recent and _REFERRING_RE.search(lowered):
        return recent_ads[-1].id
    return None


def find_annotations(history_text: str) -> List[str]:
    """Extract ad ids from annotations in stored history."""
    return [m.group("ad_id").strip()
            for m in _ANNOTATION_RE.finditer(history_text)]
